File: benchmarks/c/04_quicksort.c
// QuickSort Benchmark - C Implementation
// Array sorting algorithm with iterative quicksort

#include <stdio.h>

void swap(long* a, long* b) {
    long temp = *a;
    *a = *b;
    *b = temp;
}

int partition(long arr[], int low, int high) {
    long pivot = arr[high];
    int i = low;
    
    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            swap(&arr[i], &arr[j]);
            i++;
        }
    }
    swap(&arr[i], &arr[high]);
    return i;
}

void quicksort(long arr[], int low, int high) {
    // Use stack to avoid recursion
    int stack[100];
    int top = 0;
    
    if (low >= high) {
        return;
    }
    
    stack[top++] = low;
    stack[top++] = high;
    
    while (top > 0) {
        int h = stack[--top];
        int l = stack[--top];
        
        if (l < h) {
            int p = partition(arr, l, h);
            
            if (p > l) {
                stack[top++] = l;
                stack[top++] = p - 1;
            }
            
            if (p + 1 < h) {
                stack[top++] = p + 1;
                stack[top++] = h;
            }
        }
    }
}

int main() {
    // Benchmark parameter - array size
    int n = 100;
    long arr[100];
    
    // Initialize array with pseudo-random values
    long seed = 12345;
    for (int i = 0; i < n; i++) {
        seed = (seed * 1103515245 + 12345) % 2147483648;
        arr[i] = seed % 1000;
    }
    
    // Sort array
    quicksort(arr, 0, n - 1);
    
    // Calculate checksum to verify sort
    long checksum = 0;
    for (int i = 0; i < n; i++) {
        checksum += arr[i];
    }
    
    printf("quicksort %d elements checksum: %ld\n", n, checksum);
    return 0;
}

File: benchmarks/rust/04_quicksort.rs
// QuickSort Benchmark - Rust Implementation
// Array sorting algorithm with iterative quicksort

fn partition(arr: &mut [i64], low: usize, high: usize) -> usize {
    let pivot = arr[high];
    let mut i = low;
    
    for j in low..high {
        if arr[j] < pivot {
            arr.swap(i, j);
            i += 1;
        }
    }
    arr.swap(i, high);
    i
}

fn quicksort(arr: &mut [i64], mut low: usize, mut high: usize) {
    // Use stack to avoid recursion
    let mut stack = [0usize; 100];
    let mut top = 0;
    
    if low >= high {
        return;
    }
    
    stack[top] = low;
    top += 1;
    stack[top] = high;
    top += 1;
    
    while top > 0 {
        top -= 1;
        let h = stack[top];
        top -= 1;
        let l = stack[top];
        
        if l < h {
            let p = partition(arr, l, h);
            
            if p > l {
                stack[top] = l;
                top += 1;
                stack[top] = p - 1;
                top += 1;
            }
            
            if p + 1 < h {
                stack[top] = p + 1;
                top += 1;
                stack[top] = h;
                top += 1;
            }
        }
    }
}

fn main() {
    // Benchmark parameter - array size
    let n = 100;
    
    // Initialize array with pseudo-random values
    let mut arr = vec![0i64; n];
    let mut seed = 12345i64;
    
    for i in 0..n {
        seed = (seed * 1103515245 + 12345) % 2147483648;
        arr[i] = seed % 1000;
    }
    
    // Sort array
    quicksort(&mut arr, 0, n - 1);
    
    // Calculate checksum to verify sort
    let checksum: i64 = arr.iter().sum();
    println!("quicksort {} elements checksum: {}", n, checksum);
}
